# Session 15 - ANOVA (Analysis of Variance)


In [1]:
import pandas as pd
import numpy as np
from scipy.stats import f_oneway

## Task 2 — One-Way ANOVA on purchase_value Across Age Groups

In [2]:
df = pd.read_csv('purchase_data.csv')
print(df)

groups = [g['purchase_value'].values for _, g in df.groupby('age_group')]
f_stat, p_value = f_oneway(*groups)

print('\nF-statistic:', f_stat)
print('p-value:', p_value)

    customer_id age_group  purchase_value
0             1     18-25             850
1             2     18-25             920
2             3     18-25             780
3             4     18-25            1050
4             5     18-25             990
5             6     26-35            2100
6             7     26-35            1950
7             8     26-35            2300
8             9     26-35            1800
9            10     26-35            2050
10           11     36-50            1600
11           12     36-50            1450
12           13     36-50            1750
13           14     36-50            1550
14           15     36-50            1700

F-statistic: 80.01698018977875
p-value: 1.151867042756999e-07


## Task 3 — Interpretation

**Answer:** With F ≈ 80.0 and p ≈ 1.15e-07 (far below α = 0.05), there **is** a statistically significant difference in average purchase value across the three age groups. The 26-35 group's average purchase value stands out as clearly higher than the other two, and that gap is too large to be explained by random sampling variation alone.

## Task 4 — Between-Group and Within-Group Variance

Between-group variance is calculated as the mean square between groups (**MS_between** = SS_between / (k-1)); within-group variance is the mean square within groups (**MS_within** = SS_within / (N-k)). Their ratio is exactly the F-statistic from Task 2, which is a useful cross-check.

In [3]:
overall_mean = df['purchase_value'].mean()
k = df['age_group'].nunique()
N = len(df)

group_stats = df.groupby('age_group')['purchase_value']

ss_between = sum(len(g) * (g.mean() - overall_mean) ** 2 for _, g in group_stats)
ss_within = sum(((g - g.mean()) ** 2).sum() for _, g in group_stats)

df_between = k - 1
df_within = N - k

between_group_variance = ss_between / df_between
within_group_variance = ss_within / df_within

print('Between-group variance (MS_between):', between_group_variance)
print('Within-group variance (MS_within):', within_group_variance)
print('F cross-check (between/within):', between_group_variance / within_group_variance)

Between-group variance (MS_between): 1602206.6666666667
Within-group variance (MS_within): 20023.333333333332
F cross-check (between/within): 80.01698018977861


**What each represents:** the **between-group variance** (~1,602,207) measures how much the three age groups' *average* purchase values differ from the overall average — a big number here means the group means are spread far apart. The **within-group variance** (~20,023) measures how much purchase values vary *inside* each age group around their own group's mean — a small number here means customers within the same age group spend fairly similarly. ANOVA compares these two: when between-group variance dwarfs within-group variance (as it does here, by a factor of ~80), it's strong evidence the group means are genuinely different, not just randomly scattered.

## Task 5 — AI-Generated ANOVA Code: Zomato Customer Segments

**Prompt used:** *'Write Python code using scipy.stats to perform a one-way ANOVA comparing average monthly order amounts across three Zomato customer segments: Students, Working Professionals, and Retirees.'*

**Code (as generated, run below on dummy monthly order-amount data for each segment):**

In [4]:
students = [420, 380, 450, 400, 390, 410, 370, 430]
working_professionals = [1200, 1350, 1100, 1280, 1420, 1150, 1300, 1250]
retirees = [700, 650, 720, 680, 690, 710, 660, 705]

f_stat, p_value = f_oneway(students, working_professionals, retirees)

print('Students mean:', np.mean(students))
print('Working Professionals mean:', np.mean(working_professionals))
print('Retirees mean:', np.mean(retirees))
print('\nF-statistic:', f_stat)
print('p-value:', p_value)

Students mean: 406.25
Working Professionals mean: 1256.25
Retirees mean: 689.375

F-statistic: 365.1199289417415
p-value: 4.871100222264765e-17


**Output interpretation:** F ≈ 365.1, p ≈ 4.87e-17 — extremely significant. Average monthly order amounts clearly differ across the three segments (Working Professionals ~₹1256, Retirees ~₹689, Students ~₹406), consistent with each group having distinctly different spending capacity and ordering habits on the platform.